### PR TITLE
Update example file `imdb_lstm.py` to use dropout in RNNs

### PR DESCRIPTION
**Pull request by yaringal**

### CODE REVIEW

**Comment by PiranjaF** [comment]
Are you sure that it makes sense to add the naive dropout after the LSTM layer now that the recurrent dropout is applied? That is not obvious to me from the recurrent dropout paper.

**Comment by yaringal** [comment]
Yeah, the top dropout layer is applied to the last state (so it's only sampled once per sequence in effect). TimeDistributedDense on the other hand would be applied to each state in the sequence and indeed should have a similar dropout implementation to that of LSTM.

**Comment by PiranjaF** [comment]
Ah okay, that makes sense. Thanks!

**Comment by PiranjaF** [comment]
If you want to follow the paper more closely, I guess you should also weight decay. This would also be nice to show in an example as it's important, but only described in the documentation (not in the examples afaik). One question regarding that.. should only the LSTM layer or both LSTM + Embedding layers have this regularization?

**Comment by yaringal** [comment]
In my experiments I found that it can be very important to add regularisation over the embeddings (without dropout over the embeddings the loss seemed to decrease and increase in a snake like shape, which was resolved by adding embedding dropout).

**Comment by PiranjaF** [comment]
But would you also add weight decay to the Embedding? It's unclear to me from the paper whether the weight decay is added only to the LSTM weights or also the Embedding weights.

**Comment by yaringal** [comment]
Yeah, for the exps in the paper I used weight decay for the embedding layer as well (equivalent to a placing a prior over the linear transformation parameters). Have a look here: https://github.com/yaringal/BayesianRNN/blob/master/Sentiment_analysis_code/sentiment_lstm_regression.py#L124

**Comment by PiranjaF** [comment]
Thanks! It's important work - LSTMs really have a tendency to overfit.

Should there be any issues with setting these parameters directly from a loaded model (i.e. `model.nodes['forward'].dropout_U = 0.5` instead of setting them when defining the layers? I cannot get it to work this way, but I can't see why this approach should be problematic with the code. I compile the loaded model after setting these parameters.

Even with recurrent dropout set to 0.99 the model overfits dramatically on my dataset of 15000 sentences. Do you have a clue to why I cannot simply set the parameters directly?

**Comment by yaringal** [comment]
not sure what's going on. dropout = 1 should set model output to be constant... (completely ignore the input)

**Comment by PiranjaF** [comment]
I've started a new issue, where I've modified `imdb_lstm.py` to demonstrate the issue. It would be great if you could take a look at it! (https://github.com/fchollet/keras/issues/2259)

**Comment by PiranjaF** [comment]
I know that with recurrent dropout more iterations are needed before convergence (as stated in the paper), but it is surprising to me that each iteration also takes longer. That is not clear to me from the code.

On sentiment prediction using GPU with batch size 128 I see a 50% slowdown if I have dropout_U = 0.1, dropout_W = 0.1 for LSTM layers and dropout = 0.1 for the embedding layer. Why is that? Could it potentially be fixed?

**Comment by yaringal** [comment]
This was fixed in https://github.com/fchollet/keras/pull/2523
